### PR TITLE
 Fix #5895: REPL autocompletion crashes in certain cases

### LINKED ISSUE #5895: Completing in front of the opaque type definition will cause the REPL to crash
When I press the `<TAB>` key in front of `opaque type`, the REPL crashes: 
```scala
scala> opaque type T = Int
Exception in thread "main" java.lang.AssertionError: assertion failed: position not set for T.T # 23565 of class dotty.tools.dotc.ast.Trees$Select in <typecheck>
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:19)
	at dotty.tools.dotc.typer.Typer$.assertPositioned(Typer.scala:60)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2064)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2080)
	at dotty.tools.dotc.typer.Namer.typedAheadType$$anonfun$1(Namer.scala:1092)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:1085)
	at dotty.tools.dotc.typer.Namer.typedAheadType(Namer.scala:1092)
	at dotty.tools.dotc.typer.Namer.typeDefSig(Namer.scala:1360)
	at dotty.tools.dotc.typer.Namer$TypeDefCompleter.typeSig(Namer.scala:918)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:882)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:796)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:236)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:180)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:182)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:342)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:1932)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1957)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2033)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2068)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2066)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2080)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2099)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2141)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:698)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:701)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1987)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2034)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2068)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2080)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1095)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:1085)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1095)
	at dotty.tools.dotc.typer.Namer.rhsType$2(Namer.scala:1218)
	at dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1230)
	at dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1231)
	at dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1249)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1257)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:766)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:882)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:796)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:236)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:180)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:182)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:342)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:1932)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1957)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2033)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2068)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2080)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2099)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2141)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1674)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1970)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2033)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2068)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2080)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2099)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2141)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1795)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2012)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2034)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2068)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2080)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2152)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:61)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:35)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:65)
	at dotty.tools.repl.REPLFrontEnd.runOn(ReplFrontEnd.scala:23)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:158)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:170)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:178)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:102)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:185)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:130)
	at dotty.tools.repl.ReplCompiler.typeCheck$$anonfun$1(ReplCompiler.scala:280)
	at scala.util.Either.flatMap(Either.scala:341)
	at dotty.tools.repl.ReplCompiler.typeCheck(ReplCompiler.scala:285)
	at dotty.tools.repl.ReplDriver.completions(ReplDriver.scala:166)
	at dotty.tools.repl.ReplDriver.$anonfun$1(ReplDriver.scala:110)
	at org.jline.reader.impl.LineReaderImpl.doComplete(LineReaderImpl.java:3962)
	at org.jline.reader.impl.LineReaderImpl.expandOrComplete(LineReaderImpl.java:3875)
	at org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:585)
	at org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:404)
	at dotty.tools.repl.JLineTerminal.readLine(JLineTerminal.scala:68)
	at dotty.tools.repl.ReplDriver.readLine$1(ReplDriver.scala:115)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:125)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:130)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput$$anonfun$2$$anonfun$1(ReplDriver.scala:143)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withErr(Console.scala:196)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput$$anonfun$1(ReplDriver.scala:143)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withOut(Console.scala:167)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:143)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:130)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
```
Trying to complete in front of other definitions won't crash the REPL.

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**
# Crashes
The crashes happen due to the fact that in certain cases,
certain trees get mispositioned during code completion. E.g.
this happens with `opaque type T = Int` or `object Foo { type T = Int`.
The tree spans are set incorrectly because repl doesn't set the
sources of these trees correctly. Precisely, when typechecking
on code completion, a virtual source is used. However, when
parsing for autocompletion, a line source is used which is
created by the parsing logic. This commit makes sure that REPL
is consistent in its source choices when computing code completions.

# Completion
The above commit does not bring the code completion for opaque types, however.

REPL code completion computes completion candidates for a given
expression after typechecking it. Typechecking opaque types
desugars them from `<mods> type T = Int` to `<mods> type T = T.T`.
The completion relies on the position of the user's cursor
in the input expression to know what part of this expression to complete.
Since the cursor does not take the semantic info computed by typechecking,
the cursor points to `T.T` and not to `Int` on completion request.
Completion on synthetic `T.T` does not pick any candidates, which is
unexpected to the user. The expected behavior is to complete on `Int`,
the underlying type. This commit achieves the expected behavior by
making sure opaque types are treated as ordinary types on code completion.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala [hunk 1]
```diff
@@ -144,6 +144,7 @@ class ScalaSettings extends Settings.SettingGroup {
   val YcheckAllPatmat: Setting[Boolean] = BooleanSetting("-Ycheck-all-patmat", "Check exhaustivity and redundancy of all pattern matching (used for testing the algorithm)")
   val YretainTrees: Setting[Boolean] = BooleanSetting("-Yretain-trees", "Retain trees for top-level classes, accessible from ClassSymbol#tree")
   val YshowTreeIds: Setting[Boolean] = BooleanSetting("-Yshow-tree-ids", "Uniquely tag all tree nodes in debugging output.")
+  val YnoOpaqueTypes: Setting[Boolean] = BooleanSetting("-Yno-opaque-types", "Opaque types are treated as ordinary transparent types.")
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: I'm not sure what the underlying issue exactly is here (perhaps we just need to set better positions somewhere ?) but this is almost surely the wrong approach, we should avoid adding extra code path which are only used in very specific cases, every one of them doubles the number of possible states we need to test.

**Comment by anatoliykmetyuk** [comment]
The corner case where this solution misbehaves:

```scala
opaque type T = String; val x: T = ???; x.len<TAB>
```

The above will autocomplete to `length`, though shouldn't. The below will also autocomplete:

```scala
scala> object Foo {
     |   opaque type T = String
     |   val x: T = ???
     |   x.len<TAB>
```

If evaluated, the behaviour is correct:

```scala
scala> object Foo {
     |   opaque type T = String
     |   val x: T = ???
     |   x.length
     | }
4 |  x.length
  |  ^^^^^^^^
  |  value length is not a member of Foo.T
```

An alternative approach would be to map the cursor position to the trees generated after desugaring and typecheck. However, FTTB it seems the cursor is [oblivious](https://github.com/lampepfl/dotty/blob/8e1277180fb43523272ce182225c4762bd294bc1/compiler/src/dotty/tools/repl/ReplDriver.scala#L110) to the transformations done to the input and depends only on the input string.

File: compiler/src/dotty/tools/repl/ParseResult.scala [hunk 2]
```diff
@@ -116,7 +116,7 @@ object ParseResult {
   }
 
   /** Extract a `ParseResult` from the string `sourceCode` */
-  def apply(sourceCode: String)(implicit state: State): ParseResult =
+  def apply(sourceCode: String, sourceFile: Option[SourceFile] = None)(implicit state: State): ParseResult =
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/repl/ParseResult.scala]: We should not pass both the `sourceCode` and the `sourceFile` as the source file should contain the source code. 

You probably need:

```scala
def apply(sourceCode: String)(implicit state: State): ParseResult = 
 apply(SourceFile.virtual(str.REPL_SESSION_LINE + (state.objectIndex + 1), sourceCode))

def apply(sourceFile: SourceFile)(implicit state: State): ParseResult = {
  ...
}
```

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala (hunk 1 above)
Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: Please see my reasoning and motivation in [this comment](https://github.com/lampepfl/dotty/pull/6414#issuecomment-488304296). I agree that it is a suboptimal approach, but it seems that we don't have the info required to target the underlying type properly on completion. Do you have any thoughts on how we could do so?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: This also leaks the opaque type
```
> opaque type T = Int => Int; (null: T).app
```
outocompletes the `apply` when it should not

**Comment by smarter** [comment]
>      |   x.len<TAB>

Don't know what's going on but at this point we should be looking at the type of `x` which should be `T`, and then getting the members of `T`, since `T` is opaque this shouldn't return length.

**Comment by anatoliykmetyuk** [comment]
Ok, I removed the completion part but left the crash part since it affects not only opaque types (see the test). Hope you don't mind a bit of FP inside :) I've also created an issue #6415 to log the problem with opaque types completion.

File: compiler/src/dotty/tools/repl/ParseResult.scala [hunk 3]
```diff
@@ -116,7 +116,8 @@ object ParseResult {
   }
 
   /** Extract a `ParseResult` from the string `sourceCode` */
```
Comment (smarter) [inline compiler/src/dotty/tools/repl/ParseResult.scala]: Comment is no longer up-to-date.

File: compiler/src/dotty/tools/repl/ParseResult.scala [hunk 4]
```diff
@@ -116,7 +116,8 @@ object ParseResult {
...
-  def apply(sourceCode: String)(implicit state: State): ParseResult =
+  def apply(sourceCodeOrFile: Either[String, SourceFile])(implicit state: State): ParseResult = {
```
Comment (smarter) [inline compiler/src/dotty/tools/repl/ParseResult.scala]: I find Nicolas' suggestion in https://github.com/lampepfl/dotty/pull/6414#discussion_r280096548 clearer than wrapping things in an Either just to rewrap it later.

Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/repl/ParseResult.scala]: We have the versions for each individual types below. `Either` is to share logic.

Comment (smarter) [reply compiler/src/dotty/tools/repl/ParseResult.scala]: I get that, but if we always end up wrapping things in a SourceFile, then it makes sense to do that as soon as possible so that we don't need to think about both cases inside the method with all the logic.

Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/repl/ParseResult.scala]: Ok, good idea

File: compiler/src/dotty/tools/repl/ParseResult.scala [hunk 5]
```diff
@@ -132,8 +132,6 @@ object ParseResult {
       case _ =>
         implicit val ctx: Context = state.context
 
-        val source = SourceFile.virtual(str.REPL_SESSION_LINE + (state.objectIndex + 1), sourceCode)
-
         val reporter = newStoreReporter
         val stats = parseStats(sourceCode)(state.context.fresh.setReporter(reporter).withSource(source))
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/repl/ParseResult.scala]: The parameter `sourceCode` is never used inside `parseStats`, you could remove it

Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/repl/ParseResult.scala]: Nice catch, removed it
